sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/HadoopFileDataObject.scala

```diff
@@ -81,16 +82,32 @@ private[smartdatalake] trait HadoopFileDataObject extends FileRefDataObject with
    * Check if the input files exist.
    * Note that hadoopDir can be a specific file or a directory.
    */
-  private[smartdatalake] def checkFilesExisting(implicit context: ActionPipelineContext): Boolean = {
+  def checkFilesExisting(implicit context: ActionPipelineContext): Boolean = {
     val status = try {
       filesystem.getFileStatus(hadoopPath)
     } catch {
       case _: FileNotFoundException => return false
     }
-    status.isFile || {
-      val globPath = if (partitions.nonEmpty) new Path(hadoopPath, PartitionValues(Map()).getPartitionString(partitionLayout().get)) else hadoopPath
-      status.isDirectory && filesystem.globStatus(new Path(globPath, fileName)).exists(_.isFile)
-    }
+    status.isFile || (status.isDirectory && listDataFiles().nonEmpty)
+  }
+
+  def listDataFiles(pv: PartitionValues = PartitionValues(Map()))(implicit context: ActionPipelineContext): Iterator[Path] = {
+    val pathPattern = if (partitions.nonEmpty) new GlobPattern(new Path(new Path(hadoopPath, pv.getPartitionString(partitionLayout().get)), fileName).toString)
+    else new GlobPattern(new Path(hadoopPath, fileName).toString)
+    RemoteIteratorWrapper(filesystem.listFiles(hadoopPath, partitions.nonEmpty /*recursive*/))
+      .filter(_.isFile)
+      .map(_.getPath)
+      .filter(p => pathPattern.matches(p.toString))
+  }
+
+  def listPartitionPathsStatus(pv: PartitionValues = PartitionValues(Map()), partitionLayoutParam: String = partitionLayout().get)(implicit context: ActionPipelineContext): Seq[FileStatus] = {
```

Comment:
we could make 3 methods private. Do we keep them public on purpose to be able to extend them?

Reply:
True. On the other side i often struggle with methods not being accessible when debugging or creating custom implementations. I'm no longer a fan of making everything private/protected where possible.
In Python real private methods dont exist, they are just marked by a prefix "__". We could learn from them ;-)